File: p3.cpp
#include <iostream>
#include <fstream>
#include <cctype>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

unordered_set<string> keywords = {
    "int", "char", "float", "double", "if", "else", "while", "for",
    "return", "void", "break", "continue", "long", "short", "signed",
    "unsigned", "struct", "union", "typedef", "static"
};

vector<string> symbolTable;
vector<pair<int, string>> lexicalErrors;

bool isKeyword(const string &s) {
    return keywords.count(s);
}

bool isOperator(char c) {
    string ops = "+-*/=%<>!";
    return ops.find(c) != string::npos;
}

bool isPunctuation(char c) {
    string punc = "();{},[]";
    return punc.find(c) != string::npos;
}

bool existsInSymbolTable(const string &id) {
    for (auto &s : symbolTable)
        if (s == id) return true;
    return false;
}

int main() {
    ifstream file("input.c");
    if (!file) {
        cout << "Error opening file.\n";
        return 1;
    }

    string line;
    int lineNo = 0;
    int tokenCount = 0;
    bool inBlockComment = false;

    cout << "Tokens\n";

    while (getline(file, line)) {
        lineNo++;
        int i = 0;

        while (i < line.length()) {

            if (inBlockComment) {
                if (line[i] == '*' && i + 1 < line.length() && line[i + 1] == '/') {
                    inBlockComment = false;
                    i += 2;
                } else i++;
                continue;
            }

            if (line[i] == '/' && i + 1 < line.length() && line[i + 1] == '/') {
                break;
            }

            if (line[i] == '/' && i + 1 < line.length() && line[i + 1] == '*') {
                inBlockComment = true;
                i += 2;
                continue;
            }

            if (isspace(line[i])) {
                i++;
                continue;
            }

            if (isalpha(line[i]) || line[i] == '_') {
                string word;
                while (i < line.length() && (isalnum(line[i]) || line[i] == '_'))
                    word += line[i++];

                if (isKeyword(word))
                    cout << "Keyword: " << word << endl;
                else {
                    cout << "Identifier: " << word << endl;
                    if (!existsInSymbolTable(word))
                        symbolTable.push_back(word);
                }
                tokenCount++;
            }

            else if (isdigit(line[i])) {
                string num;
                while (i < line.length() && isalnum(line[i]))
                    num += line[i++];

                bool valid = true;
                for (char c : num) {
                    if (!isdigit(c)) {
                        valid = false;
                        break;
                    }
                }

                if (valid) {
                    cout << "Constant: " << num << endl;
                    tokenCount++;
                } else {
                    lexicalErrors.push_back({lineNo, num});
                }
            }

            else if (line[i] == '\'') {
                string chConst;
                chConst += line[i++];

                if (i < line.length())
                    chConst += line[i++];

                if (i < line.length() && line[i] == '\'') {
                    chConst += line[i++];
                    cout << "Constant: " << chConst << endl;
                    tokenCount++;
                } else {
                    lexicalErrors.push_back({lineNo, chConst});
                }
            }

            else if (isOperator(line[i])) {
                cout << "Operator: " << line[i] << endl;
                tokenCount++;
                i++;
            }

            else if (isPunctuation(line[i])) {
                cout << "Punctuation: " << line[i] << endl;
                tokenCount++;
                i++;
            }

            else {
                string err(1, line[i]);
                lexicalErrors.push_back({lineNo, err});
                i++;
            }
        }
    }

    file.close();

    cout << "\nLexical Errors\n";
    for (auto &e : lexicalErrors)
        cout << "Line " << e.first << " : " << e.second << " invalid lexeme\n";

    cout << "\nSymbol table Entries\n";
    for (int i = 0; i < symbolTable.size(); i++)
        cout << i + 1 << ") " << symbolTable[i] << endl;

    cout << "\nTotal Tokens: " << tokenCount << endl;

    return 0;
}   

File: p3.java
import java.io.*;
import java.util.*;

public class p3 {
    
    static HashSet<String> keywords = new HashSet<>(Arrays.asList(
        "auto", "break", "case", "char", "const", "continue", "default", "do", "double", "else", "enum", "extern",
        "float", "for", "goto", "if", "inline", "int", "long", "register", "restrict", "return", "short", "signed",
        "sizeof", "static", "struct", "switch", "typedef", "union", "unsigned", "void", "volatile", "while", "printf", "main"
    ));
    
    static ArrayList<String> symbolTable = new ArrayList<>();
    static ArrayList<String> lexicalErrors = new ArrayList<>();
    
    static boolean isKeyword(String str) {
        return keywords.contains(str);
    }
    
    static boolean isIdentifierChar(char ch, boolean isStart) {
        return isStart ? (Character.isLetter(ch) || ch == '_') : (Character.isLetterOrDigit(ch) || ch == '_');
    }
    
    static boolean isOperator(char ch) {
        return ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '=' || ch == '%' || ch == '>' || ch == '<';
    }
    
    static boolean isPunctuation(char ch) {
        return ch == '(' || ch == ')' || ch == '{' || ch == '}' || ch == ';' || ch == ',' || ch == '"' || ch == '[' || ch == ']';
    }
    
    static void processIdentifier(String word) {
        if (isKeyword(word))
            return; // Don't add keywords to symbol table
        if (!symbolTable.contains(word))
            symbolTable.add(word);
    }
    
    public static void main(String[] args) {
        try {
            BufferedReader file = new BufferedReader(new FileReader("C:\\Users\\Vatsal\\Desktop\\College Sem 6\\Compiler Construction\\input.c"));
            
            String line;
            int tokenCount = 0;
            
            while ((line = file.readLine()) != null) {
                StringBuilder word = new StringBuilder();
                boolean inString = false;
                
                for (int i = 0; i < line.length(); i++) {
                    char ch = line.charAt(i);
                    
                    if (inString) {
                        word.append(ch);
                        if (ch == '"') {
                            inString = false;
                            System.out.println("String: " + word);
                            tokenCount++;
                            word.setLength(0);
                        }
                        continue;
                    }
                    
                    if (Character.isWhitespace(ch) || isPunctuation(ch) || isOperator(ch)) {
                        if (word.length() > 0) {
                            if (Character.isDigit(word.charAt(0))) {
                                boolean isValidConstant = true;
                                for (int j = 0; j < word.length(); j++) {
                                    char c = word.charAt(j);
                                    if (!Character.isDigit(c) && c != '.' && c != 'e' && c != '-') {
                                        isValidConstant = false;
                                        break;
                                    }
                                }
                                if (!isValidConstant) {
                                    lexicalErrors.add(word.toString());
                                } else {
                                    System.out.println("Constant: " + word);
                                    tokenCount++;
                                }
                            } else if (isIdentifierChar(word.charAt(0), true)) {
                                processIdentifier(word.toString());
                                System.out.println("Identifier: " + word);
                                tokenCount++;
                            }
                            word.setLength(0);
                        }
                        
                        if (ch == '"') {
                            inString = true;
                            word.append(ch);
                        } else if (isOperator(ch)) {
                            System.out.println("Operator: " + ch);
                            tokenCount++;
                        } else if (isPunctuation(ch)) {
                            System.out.println("Punctuation: " + ch);
                            tokenCount++;
                        }
                    } else {
                        word.append(ch);
                    }
                }
                
                if (word.length() > 0 && !inString) {
                    if (Character.isDigit(word.charAt(0))) {
                        boolean isValidConstant = true;
                        for (int j = 0; j < word.length(); j++) {
                            char c = word.charAt(j);
                            if (!Character.isDigit(c) && c != '.' && c != 'e' && c != '-') {
                                isValidConstant = false;
                                break;
                            }
                        }
                        if (!isValidConstant) {
                            lexicalErrors.add(word.toString());
                        } else {
                            System.out.println("Constant: " + word);
                            tokenCount++;
                        }
                    } else if (isIdentifierChar(word.charAt(0), true)) {
                        processIdentifier(word.toString());
                        System.out.println("Identifier: " + word);
                        tokenCount++;
                    }
                }
            }
            
            file.close();
            
            System.out.println("\nTotal Tokens: " + tokenCount);
            
            System.out.println("\nSymbol Table Entries:");
            for (String symbol : symbolTable)
                System.out.println(symbol);
            
            System.out.println("\nLexical Errors:");
            for (String error : lexicalErrors)
                System.out.println(error);
                
        } catch (IOException e) {
            System.out.println("Failed to open input file.");
            e.printStackTrace();
        }
    }
}